frontend/src/components/GameCanvas.jsx

```diff
@@ -6,48 +6,60 @@ import { OrbitControls, PerspectiveCamera, Text } from '@react-three/drei';
 import { levels } from "../levels";
 
 import Castle from "../entities/Castle";
-import MapView from "./MapView";
+import MapView from "./views/MapView";
 import Portal from "../entities/Portal";
-import EnemyView from "./EnemyView";
-import StructureView from "./StructureView";
-import ProjectileView from "./ProjectileView";
+import EnemyView from "./views/EnemyView";
+import StructureView from "./views/StructureView";
+import ProjectileView from "./views/ProjectileView";
 import ArrowTower from "../entities/towers/ArrowTower";
 import RockTower from "../entities/towers/RockTower";
-import GameInfo from "./GameInfo";
+import GameInfo from "./ui/GameInfo";
+import BuyMenu from "./ui/BuyMenu";
+import Guy from "../entities/enemies/Guy";
+import Arrow from "../entities/projectiles/Arrow";
+import Rock from "../entities/projectiles/Rock";
 
 export default function GameCanvas({game}) {
     const { gameMap } = game;
     const { width, depth, height, heightMap } = gameMap;
     const [castle, portal] = [new Castle(width-1, depth-1, heightMap.at(-1).at(-1)), new Portal(0, 0, heightMap[0][0])];
   
+    const towerConstructors = [ArrowTower, RockTower];
+    const [newTower, setNewTower] = useState(null);
+
+    const [goldReward, setGoldReward] = useState(100);
     const [enemies, setEnemies] = useState([]);
-    const [towers, setTowers] = useState([
+    const [structures, setStructures] = useState([
         castle,
         portal,
         new ArrowTower(width-1, 0, heightMap.at(-1).at(0)),
         new RockTower(0, depth-1, heightMap.at(0).at(-1))
     ]);
-    game.towers = towers; // will need to change to game.addTower(tower) when building towers is implemented
     
     const [projectiles, setProjectiles] = useState([]);
     const ref = useRef(null);
     
     useEffect(() => {
+        // preload models
+        setEnemies([new Guy(0, 0, -1, 0.001)]);
+        setProjectiles([new Arrow(0, 0, -1, 0.001), new Rock(0, 0, -1, 0.001)]);
+        
+        for (let structure of structures) game.addTower(structure);
+
         const animate = () => {
             game.tick()
 
             handleEnemiesAtCastle(game.enemies);
-            setTowers([...game.towers]);
+            towersAttack(game.enemies);
 
             setEnemies([...game.enemies]);
-            towersAttack(game.enemies);
             setProjectiles([...game.projectiles]);
 
             checkLevelOver();
 
             ref.current = requestAnimationFrame(animate);
         }
-        ref.current = requestAnimationFrame(animate);
+        setTimeout(() => ref.current = requestAnimationFrame(animate), 100);
```

Comment:
I may have missed this before, why are we calling `requestAnimationFrame` directly? I believe react three fiber is supposed to take care of that for us. Maybe `useFrame` is what we need for the game loop instead.

Reply:
Probably worth tackling this one same time as the freezing issue. We can save it for a future PR.

Reply:
Yeah, sounds good. The freezing seems to have been cleaned up by preloading the models though, fyi